iceoryx_hoofs/include/iceoryx_hoofs/error_handling/error_handling.hpp

```diff
@@ -276,9 +245,95 @@ class ErrorHandler
 /// errorHandler(Error::kTEST__ASSERT_CALLED);
 /// ASSERT_TRUE(called);
 /// @endcode
+/// @tparam[in] Error type of the enum which is used to report the error
+/// @todo #590 use enable_if with is_enum
+template <typename Error>
 void errorHandler(const Error error,
                   const std::function<void()>& errorCallBack = std::function<void()>(),
                   const ErrorLevel level = ErrorLevel::FATAL) noexcept;
+
+template <typename Error>
+using TypedHandlerFunction = std::function<void(const Error, const ErrorLevel)>;
+
+using HandlerFunction = std::function<void(const uint32_t, const char*, const ErrorLevel)>;
+
+/// @brief This handler is needed for unit testing, special debugging cases and
+///         other corner cases where we'd like to explicitly suppress the
+///         error handling.
+class ErrorHandler
+{
+    template <typename Error>
+    friend void
+    errorHandler(const Error error, const std::function<void()>& errorCallBack, const ErrorLevel level) noexcept;
+
+  public:
+    template <typename Error>
+    static cxx::GenericRAII setTemporaryErrorHandler(const TypedHandlerFunction<Error>& newHandler) noexcept;
+
+  protected:
+    static void reactOnErrorLevel(const ErrorLevel level, const char* errorText) noexcept;
+
+  private:
+    static void
+    defaultHandler(const uint32_t error, const char* errorName, const ErrorLevel level = ErrorLevel::FATAL) noexcept;
+
+    static iox::HandlerFunction handler;
+    /// Needed, if you want to exchange the handler. Remember the old one and call it if it is not your error. The error
+    /// mock needs to be the last one exchanging the handler in tests.
+    static std::mutex handler_mutex;
+};
+
+const char* toString(const Error error) noexcept;
+
+template <typename Error>
+cxx::optional<iox::TypedHandlerFunction<Error>> typedHandler;
+
+
+/// @todo #590 move the implementations below to .inl
+template <typename Error>
+inline void errorHandler(const Error error,
+                         const std::function<void()>& errorCallBack IOX_MAYBE_UNUSED,
```

Comment:
The removal will be done either in #1028 or a follow-up.

Reply:
Maybe we should create small separate issues for the error handler since I am already a little bit lost. And you have several comments where you state, will be done in this PR or will be done in another follow up  or in this issue and so on. And when I look at those PRs, issues etc. either the task is done incompletely or nothing is mentioned in the issue to this particular task.

For me as a reviewer it is hard to track all of this.

Reply:
I'll close this PR and will open an new one based on #1099